graphbolt/src/cuda/extension/gpu_graph_cache.h

```diff
@@ -0,0 +1,81 @@
+/**
+ *   Copyright (c) 2023, GT-TDAlab (Muhammed Fatih Balin & Umit V. Catalyurek)
+ *   All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ * @file cuda/gpu_graph_cache.h
+ * @brief Header file of GPU graph cache.
+ */
+
+#ifndef GRAPHBOLT_GPU_GRAPH_CACHE_H_
+#define GRAPHBOLT_GPU_GRAPH_CACHE_H_
+
+#include <torch/custom_class.h>
+#include <torch/torch.h>
+
+#include <limits>
+#include <type_traits>
+
+namespace graphbolt {
+namespace cuda {
+
+class GpuGraphCache : public torch::CustomClassHolder {
+  static constexpr double load_factor = 0.8;
+  static constexpr int growth_factor = 2;
+
+ public:
+  /**
+   * @brief Constructor for the GpuGraphCache struct.
+   *
+   * @param num_nodes The node capacity of GPU cache.
+   * @param num_edges The edge capacity of GPU cache.
+   * @param dtype The node id datatype.
+   */
+  GpuGraphCache(
+      const int64_t num_edges, const int64_t threshold,
+      torch::ScalarType indptr_dtype, std::vector<torch::ScalarType> dtypes);
```

Comment:
why the dtypes are vector?

Reply:
So that we can cache a different number of edge tensors, such as indices, type_per_edge and probs_or_mask.